coord: split up advance table processing

mjibson opened the pull request:
Previously, all tables were advanced in the same append call. The
implementation at the bottom of that was serial, so coord would be
blocked until each table was processed.

Break this up so that each table is processed one at a time at the lowest
message priority. This means that incoming user queries are blocked for
at most the time it takes to process a single table advance instead of
all of them.

Use a data structure that allows for the 1s advance loop to trigger faster
than it takes us to advance all of the tables. This could happen if:

- a user has many tables
- there are many mz instances all using the same postgres instance and its load increases query times
- the postgres instance hangs for a few seconds
- network or other delays somewhere in the stack

This solution is much simpler than attempting to split off the storage
controller into its own task.

### Motivation

  * This PR fixes a recognized bug. Informs #12944

### Testing

- [ ] This PR has adequate test coverage / QA involvement has been duly considered.

### Release notes

This PR includes the following [user-facing behavior changes](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/guide-changes.md#what-changes-require-a-release-note):

  - n/a

jkosh44 commented:
I thought I read somewhere today on slack that persist/storage was able to move the table advancements to happen in parallel? Either way this will probably still help lower the load on the stash implementation depending on how it's implemented.

jkosh44 commented on src/coord/src/coord.rs:
```diff
@@ -332,6 +334,49 @@ fn concretize_replica_config(
 
+/// Holds tables needing advancement.
+struct AdvanceTables<T> {
+    map: HashMap<GlobalId, T>,
```
Is there ever a scenario where the values would not all be the same timestamp? If not it might make sense to change this to have a `Set<GlobalId>` and single `T` field.

mjibson commented:
https://materializeinc.slack.com/archives/C02KE368L8P/p1654622518724229?thread_ts=1654557605.225829&cid=C02KE368L8P caused me to think that it's not parallel and isn't going to be anytime soon?

guswynn commented on src/coord/src/coord.rs:
```diff
@@ -332,6 +334,49 @@ fn concretize_replica_config(
+        // some things never get visited in busy coordinators.
+        let key: Option<GlobalId> = self.map.keys().next().cloned();
+        match key {
+            None => futures::future::pending().await,
```
I would put a comment on this function explaining that `recv` will never finish if there is nothing to do, and that this is required by the current coord implementation and nothing else

mjibson replied:
The never finish comment already exists. I was going to add some words talking about its use in coord but I thought it was clearer just to not instead?

jkosh44 commented on src/coord/src/coord.rs:
```diff
@@ -332,6 +334,49 @@ fn concretize_replica_config(
+    }
+
+    // Returns the next table to advance. Blocks forever if there are none.
+    async fn recv(&mut self) -> (GlobalId, T) {
```
It might be nice to be able to toggle this to return multiple `GlobalId`s at once. Only advancing a single table at once seems like the opposite extreme of advancing all tables at once. The right tradeoff might be somewhere in the middle.

jkosh44 replied:
This is subjective though, so feel free to disagree.

mjibson replied:
persist changed their stuff to pipeline. It's not parallel, but it at least stops wait for responses before sending the next request, so things can be done much more in parellel. This PR would essentially negate all of that and overall increase the total amount of time it takes to close all tables. So I do think that there is *some* number of things we should be batching here.

The problem we are attempting to address is: an incoming query is blocked by table closing. We don't know when these queries will arrive. Let's say we have some target like: incoming queries are not delayed due to table closing by more than 50ms. This function could measure the time of its previous close attempt (or some weighted avg), and either increase or decrease the number of tables it attempts to close based on how long the previous attempt took relative to the target. This has some benefits like: if the pg server gets more loaded we back off, if it gets faster we speed up, if persist is able to parallelize those requests (say, via a connection pool), then we can aggressively speed up batching. This seems worth doing, I'll look at this.

guswynn commented on src/coord/src/coord.rs:
```diff
@@ -837,6 +886,17 @@ impl<S: Append + 'static> Coordinator<S> {
+                // advance_tables is fully emptied, this allows us to replace an old request
+                // with a new one, avoiding duplication of work, which wouldn't be possible if
+                // we had already sent all AdvanceLocalInput messages on a channel.
+                m = self.advance_tables.recv() => {
+                    Message::AdvanceLocalInput(m.0, m.1)
+                },
```
```suggestion
                (id, timestamp) = self.advance_tables.recv() => {
                    Message::AdvanceLocalInput(id, timestamp)
                },
```

guswynn reviewed (commented):
I think @jkosh44 and @aljoscha will need to sign off on this, but looks good to me!

benesch commented:
This is cool. Deferring actual review to other folks though.

aljoscha commented:
> I thought I read somewhere today on slack that persist/storage was able to move the table advancements to happen in parallel? Either way this will probably still help lower the load on the stash implementation depending on how it's implemented.

We increased parallelism a bit.Tthe consensus requests are now pipelined in the client, but still not fully concurrent.

aljoscha commented on src/coord/src/coord.rs:
```diff
@@ -332,6 +334,49 @@ fn concretize_replica_config(
+
+    // Returns the next table to advance. Blocks forever if there are none.
+    async fn recv(&mut self) -> (GlobalId, T) {
+        // TODO(mjibson): In order to prevent starvation, we need to make this random
```
It's more complicated, but could we keep both a map and a queue in `AdvanceTables`? We would use the queue to ensure we work off advancement requests fairly, and we use the map to keep track of what the latest timestamp is that we need to advance to.

mjibson replied:
Yes, done.

aljoscha reviewed (commented):
Can there now be a race condition because the coordinator thinks that a table was advanced to a certain timestamp but the actual update in persist didn't yet go through?

Before, we would update our own view of where times are, in reaction to `Message::AdvanceLocalInputs`, and then we would (at the end of the loop) block on advancing the tables in persist. Now, we don't do that latter part for all tables atomically, anymore. When we now try to append an INSERT to a table, it could fail because the frontier in persist is not what we expect.

aljoscha commented:
@mjibson This is what I had in mind, when I wrote that fairness comment:

```
From 8c02092616d41de456ca5a71226b6fe29b9fdcc8 Mon Sep 17 00:00:00 2001
From: Aljoscha Krettek <aljoscha.krettek@gmail.com>
Date: Thu, 9 Jun 2022 14:45:41 +0200
Subject: [PATCH] fair table downgrades

---
 src/coord/src/coord.rs | 8 ++++----
 1 file changed, 4 insertions(+), 4 deletions(-)

diff --git a/src/coord/src/coord.rs b/src/coord/src/coord.rs
index 1e4cc523e..231b551bd 100644
--- a/src/coord/src/coord.rs
+++ b/src/coord/src/coord.rs
@@ -336,12 +336,14 @@ fn concretize_replica_config(
 /// Holds tables needing advancement.
 struct AdvanceTables<T> {
     map: HashMap<GlobalId, T>,
+    work: VecDeque<GlobalId>,
 }
 
 impl<T: CoordTimestamp> AdvanceTables<T> {
     fn new() -> Self {
         Self {
             map: HashMap::new(),
+            work: VecDeque::new(),
         }
     }
 
@@ -356,16 +358,14 @@ impl<T: CoordTimestamp> AdvanceTables<T> {
             }
             Entry::Vacant(entry) => {
                 entry.insert(ts);
+                self.work.push_back(id);
             }
         }
     }
 
     // Returns the next table to advance. Blocks forever if there are none.
     async fn recv(&mut self) -> (GlobalId, T) {
-        // TODO(mjibson): In order to prevent starvation, we need to make this random
-        // or somehow fair. As is, the iteration order is arbitrary, so it is possible
-        // some things never get visited in busy coordinators.
-        let key: Option<GlobalId> = self.map.keys().next().cloned();
+        let key: Option<GlobalId> = self.work.pop_front();
         match key {
             None => futures::future::pending().await,
             Some(id) => {
-- 
2.36.1
```

jkosh44 commented:
> Can there now be a race condition because the coordinator thinks that a table was advanced to a certain timestamp but the actual update in persist didn't yet go through?
> 
> Before, we would update our own view of where times are, in reaction to `Message::AdvanceLocalInputs`, and then we would (at the end of the loop) block on advancing the tables in persist. Now, we don't do that latter part for all tables atomically, anymore. When we now try to append an INSERT to a table, it could fail because the frontier in persist is not what we expect.

I think you're right, the specific series of events would look something like this:
1. Global timestamp starts at `ts`.
2. An `AdvanceLocalInputs` message pushes the global timestamp to `ts+1` and updates all values in the `advance_tables` to `ts+1`
3. Someone writes to table `t` which gets a timestamp of `ts+2` and pushes the global timestamp to `ts+2`.
4. An `AdvanceLocalInput` message causes us to try and advance table `t` to `ts+1` which will fail because it's already at `ts+2`.

We could have writes either increase the value in `advance_tables` or remove the key which should fix this.

I don't think an `INSERT` will ever fail because even though we may not know the actual frontier, we know an upper bound for it and the `INSERT` will always happen at a larger value.

mjibson commented:
After a refactor, the described race condition doesn't occur because there's a single advance ts that is tracked, so the insert will cause all tables to be bumped to the insert time. But a branch was added to filter out those tables just for performance.

Added the vecdeque for work fairness.

Also added a hopefully not-too-complicated batching mechanism with some gut instinct user responsiveness goals.

RFAL.

jkosh44 reviewed (approved):
LGTM
